src/parsers/manifest/metaplaylist/representation_index.ts

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright 2015 CANAL+ Group
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import IRepresentationIndex from "../../../manifest/representation_index";
+import { ISegment } from "../../../manifest/representation_index/interfaces";
+
+/**
+ * The MetaRepresentationIndex is wrapper for all kind of indexes (dash, smooth, etc)
+ *
+ * Ut wraps methods from origin indexes, while taking into account of the offset induced
+ * by metaplaylist. It makes a bridge between the metaplaylist timeline, and the original
+ * timeline of content. (e.g. the segment whose "meta" time is 1500, is actually a
+ * segment whose original time is 200, played with an offset of 1300)
+ */
+export default class MetaRepresentationIndex implements IRepresentationIndex{
+  private _wrappedIndex: IRepresentationIndex;
+  private _timeOffset: number;
+  private _transport: string;
+  private _contentEnd: number;
+
+  constructor(
+      index: IRepresentationIndex,
+      timeOffset: number,
+      transport: "dash"|"smooth",
+      contentEnd: number
+  ){
+    this._wrappedIndex = index;
+    this._timeOffset = timeOffset;
+    this._transport = transport;
+    this._contentEnd = contentEnd;
+  }
+
+  public getInitSegment() {
+    const segment = this._wrappedIndex.getInitSegment();
+    if (segment === null) {
+      return null;
+    }
+    segment.privateInfos = segment.privateInfos || {};
+    segment.privateInfos[this._transport] = null;
+    return segment;
+  }
+
+  public getSegments(up : number, duration : number) : ISegment[] {
+    return this._wrappedIndex.getSegments(
+      up - this._timeOffset,
+      duration
+    ).map((segment) => {
```

Comment:
Shouldn't you re-add timeOffset here?

Reply:
We probably should.

Basically, for me:

```
+----------------+ 1. Ask for segments infos from t to t+n  +--------------+
|                | ---------------------------------------> |              |
|                |                                          |              |
|                | <--------------------------------------- |              |
|                |  4. Gives segment infos asked (offseted) |              |
|                |                                          |              |
|      CORE      |                                          | METAPLAYLIST |
|                |                                          |              |
|                | 5. Ask to download segment (offseted)    |              |
|                | ---------------------------------------> |              |
|                |                                          |              |
|                | <--------------------------------------- |              |
+----------------+              8. Gives segment (offseted) +--------------+
                                                             | ^     |  ^
                                                             | |     |  |
                                                             | |     |  |
   +------------+   2. get segment infos from t-offset to t+n-offset |  |
   | +-------+  | <------------------------------------------+ |     |  |
   | |       |  |                                              |     |  |
   | |  HLS  |  |                                              |     |  |
   | |       |  |  3. Gives segment infos asked                |     |  |
   | +-------+  | ---------------------------------------------+     |  |
   | +-------+  |                                                    |  |
   | |       |  |                                                    |  |
   | |  HSS  |  |                                                    |  |
   | |       |  |                                                    |  |
   | +-------+  |                                                    |  |
   | +--------+ |  6. Ask to download non-offseted (normal) segments |  |
   | |        | | <--------------------------------------------------+  |
   | |  DASH  | |                                                       |
   | |        | |  7. Gives normal segment                              |
   | +--------+ | ------------------------------------------------------+
   +------------+                                                        
                                                                         
```

Which means:
  - Core -> MetaPlaylist = everything from infos to data is offseted (Core doesn't even know that it is offseted)
  - MetaPlaylist -> Net = Nothing is offseted, net believes it to be the regular segment in the manifest
  - Net -> MetaPlaylist = Nothing is offseted, net believes it to be the regular segment in the manifest
  - MetaPlaylist -> Core = everything from infos to data is offseted